### PR TITLE
Update GHA workflow

### PR DESCRIPTION
**Pull request by hmgaudecker**
#### Changes

Some stuff from [recent work](https://github.com/OpenSourceEconomics/econ-project-templates/commit/15d712a1db5d96ad665f9ff8be0c0f87733e3b03) over on econ-project-templates.

- Use checkout@v3 instead of v2 to get rid of deprecation warnings
- Use micromamba instead of conda-incubator (speed, better interface)

Have a look, maybe useful here, too!

### CODE REVIEW

**Comment by tobiasraabe** [comment]
Wow, does it reduce the runtime of the whole workflow from 1:45h to 1:15h? Compare https://github.com/pytask-dev/cookiecutter-pytask-project/actions/runs/3505768446/usage to https://github.com/pytask-dev/cookiecutter-pytask-project/actions/runs/3600786078/usage. Have you seen similar speed-ups?

**Comment by tobiasraabe** [comment]
We need to add the same to the workflow in the template.

**Comment by hmgaudecker** [comment]
> Wow, does it reduce the runtime of the whole workflow from 1:45h to 1:15h? Compare https://github.com/pytask-dev/cookiecutter-pytask-project/actions/runs/3505768446/usage to https://github.com/pytask-dev/cookiecutter-pytask-project/actions/runs/3600786078/usage. Have you seen similar speed-ups?

I have not compared in detail, also because too many things in between two working versions also in terms of the tests. But I would not be surprised based on local observations if conda requires half a minute more to resolve the environment on a slow machine.

**Comment by hmgaudecker** [comment]
One thing you may want to think about here is to get rid of tox, too (at least in the outer project). It definitely is a helpful tool when you want to make sure a library is working with all kinds of different dependencies. But here the main issue seems to be testing across operation systems, then you need GHA anyhow.

What I view as an issue is that with the current setup, you have 3 different places for conda environments
- GHA
- tox
- `environment.yml`

Same for the inner project. Left me confused while debugging things on the other machine (same here why things suddently [failed in 3af4779](https://github.com/pytask-dev/cookiecutter-pytask-project/pull/30/checks)... There I am down to one environment specified in the yaml file (plus the same for the inner project).

**Comment by tobiasraabe** [comment]
> But I would not be surprised based on local observations if conda requires half a minute more to resolve the environment on a slow machine.
>
> Oh, I was reading minutes instead of hours 😆. Still, comparing individual tests you see that you gain a minute on Ubuntu for the environment generation step, will be much more on other OS's.

I think the caching that is saving environments for one day is the main driver here. Especially on windows. And you are right, it would be even faster without tox.

**Comment by tobiasraabe** [comment]
I am also not super happy with the tox setup. Here, it prevents caching, and you mentioned another issue. I generally like isolated environments for testing, but does it have many more benefits compared to using the main environment?

**Comment by hmgaudecker** [comment]
> I think the caching that is saving environments for one day is the main driver here.

On Ubuntu, that's 10s (27 on first run vs 17 on second AFAICT).

File: hooks/post_gen_project.py [hunk 1]
```diff
@@ -60,18 +60,30 @@ def main() -> None:
         )
 
     if "{{ cookiecutter.create_conda_environment_at_finish }}" == "yes":
-        if shutil.which("mamba") is not None:
-            conda_exe = shutil.which("mamba")
+
+        if _mamba := shutil.which("mamba"):
+            conda_exe = _mamba
```
Comment (tobiasraabe) [inline hooks/post_gen_project.py:65]:
```suggestion
        if shutil.which("mamba") is not None:
            conda_exe = shutil.which("mamba")
```

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -30,17 +30,19 @@ jobs:
...
           channels: conda-forge,nodefaults
-          mamba-version: "*"
-
-      - name: Install core dependencies.
-        shell: bash -l {0}
-        run: mamba install -c conda-forge tox-conda coverage
+          extra-specs: |
+            python=${{ matrix.python-version }}
+            mamba
+            conda=22.9
```
Comment (tobiasraabe) [inline .github/workflows/main.yml:42]:
```suggestion
            conda
```

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -58,14 +60,19 @@ jobs:
...
-      - name: Install core dependencies.
-        shell: bash -l {0}
-        run: conda install -c conda-forge tox-conda
+          environment-file: false
+          environment-name: gha-docs
+          channels: conda-forge,nodefaults
+          extra-specs: |
+            python=3.11
+            mamba
+            conda=22.9
```
Comment (tobiasraabe) [inline .github/workflows/main.yml:72]:
```suggestion
```

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -58,14 +60,19 @@ jobs:
...
+            tox-conda
+            coverage
```
Comment (tobiasraabe) [inline .github/workflows/main.yml:74]:
```suggestion
```

File: {{cookiecutter.project_slug}}/.github/workflows/main.yml [hunk 5]
```diff
@@ -78,14 +80,19 @@ jobs:
...
-        run: conda install -c conda-forge tox-conda
+          environment-file: false
+          environment-name: gha-docs
+          channels: conda-forge,nodefaults
+          extra-specs: |
+            python=3.11
+            mamba
+            conda=22.9
+            tox-conda
+            coverage
```
Comment (tobiasraabe) [inline {{cookiecutter.project_slug}}/.github/workflows/main.yml:94]:
```suggestion
```

File: {{cookiecutter.project_slug}}/.github/workflows/main.yml [hunk 6]
```diff
@@ -78,14 +80,19 @@ jobs:
...
-      - name: Install core dependencies.
-        shell: bash -l {0}
-        run: conda install -c conda-forge tox-conda
+          environment-file: false
+          environment-name: gha-docs
+          channels: conda-forge,nodefaults
+          extra-specs: |
+            python=3.11
+            mamba
+            conda=22.9
```
Comment (tobiasraabe) [inline {{cookiecutter.project_slug}}/.github/workflows/main.yml:92]:
```suggestion
```

File: {{cookiecutter.project_slug}}/.github/workflows/main.yml [hunk 7]
```diff
@@ -30,17 +30,19 @@ jobs:
...
           channels: conda-forge,nodefaults
-          mamba-version: "*"
-
-      - name: Install core dependencies.
-        shell: bash -l {0}
-        run: mamba install -c conda-forge tox-conda coverage
+          extra-specs: |
+            python=${{ matrix.python-version }}
+            mamba
+            conda=22.9
```
Comment (tobiasraabe) [inline {{cookiecutter.project_slug}}/.github/workflows/main.yml:42]:
```suggestion
```

File: {{cookiecutter.project_slug}}/.github/workflows/main.yml [hunk 8]
```diff
@@ -30,17 +30,19 @@ jobs:
...
+            tox-conda
+            coverage
```
Comment (tobiasraabe) [inline {{cookiecutter.project_slug}}/.github/workflows/main.yml:44]:
```suggestion
```

File: .github/workflows/main.yml [hunk 9]
```diff
@@ -30,17 +30,18 @@ jobs:
...
-          mamba-version: "*"
-
-      - name: Install core dependencies.
-        shell: bash -l {0}
-        run: mamba install -c conda-forge tox-conda coverage
+          extra-specs: |
+            python=${{ matrix.python-version }}
+            mamba
+            tox-conda
+            coverage
```
Comment (tobiasraabe) [inline .github/workflows/main.yml:43]:
```suggestion
```

**Pull request merged**